File: app/src/main/java/com/rjw/audioprofile/service/QuickPanel.java
package com.rjw.audioprofile.service;

import android.graphics.drawable.Icon;
import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;

import com.rjw.audioprofile.utils.AudioProfileList;

public class QuickPanel extends TileService {
    @Override
    public void onCreate() {
        super.onCreate();
        setIcon();
    }

    @Override
    public void onClick() {
        super.onClick();
        int nextProfile = AudioProfileList.getCurrentProfile() + 1;
        if(nextProfile >= AudioProfileList.NO_PROFILES) {
            nextProfile = 0;
        }
        AudioProfileList.setCurrentProfile(nextProfile);
        setIcon();
    }

    @Override
    public void onTileAdded() {
        super.onTileAdded();
        AudioProfileList.initialise(getBaseContext());
    }

    @Override
    public void onStartListening() {
        super.onStartListening();
        AudioProfileList.initialise(getBaseContext());
        setIcon();
    }

    public void setIcon() {
        final Tile tile = getQsTile();
        if(tile != null) {
            try {
                final int profile = AudioProfileList.getCurrentProfile();
                final AudioProfileList.AudioProfile audioProfile = AudioProfileList.getProfile(profile);
                tile.setState(Tile.STATE_ACTIVE);
                tile.setIcon(Icon.createWithResource(getApplicationContext(), AudioProfileList.getIconResource(audioProfile.icon)));
                tile.setLabel(audioProfile.name);
                tile.updateTile();
                AudioProfileList.applyProfile(getApplicationContext());
            } catch (Exception e) {
                // Do nothing - no profiles found.
            }
        }
    }
}

File: app/src/main/java/com/rjw/audioprofile/service/QuickPanel.kt
package com.rjw.audioprofile.service

import android.graphics.drawable.Icon
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import com.rjw.audioprofile.R
import com.rjw.audioprofile.utils.AudioProfileList
import java.lang.Exception

class QuickPanel : TileService() {
    /**
     * Create the quick panel item.
     */
    override fun onCreate() {
        super.onCreate()
        setIcon()
    }

    /**
     * Handle the user clicking on the quick panel icon.
     */
    override fun onClick() {
        super.onClick()
        var nextProfile = AudioProfileList.currentProfile + 1
        if(nextProfile >= AudioProfileList.noProfiles) {
            nextProfile = 0
        }
        AudioProfileList.currentProfile = nextProfile
        AudioProfileList.profileLocked = false
        setIcon()
    }

    /**
     * Initialise the newly added tile.
     */
    override fun onTileAdded() {
        super.onTileAdded()
        AudioProfileList.initialise(baseContext)
    }

    /**
     * Start listening for user events.
     */
    override fun onStartListening() {
        super.onStartListening()
        AudioProfileList.initialise(baseContext)
        setIcon()
    }

    /**
     * Set the icon for the quick panel tile.
     */
    private fun setIcon() {
        val tile = qsTile
        if(tile != null) {
            try {
                val profile = AudioProfileList.currentProfile
                val audioProfile = AudioProfileList.getProfile(profile)
                tile.state = Tile.STATE_ACTIVE
                tile.icon = Icon.createWithResource(applicationContext, AudioProfileList.getIconResource(audioProfile.icon))
                tile.label = "${audioProfile.name} ${if(AudioProfileList.profileLocked) "\r${getString(R.string.locked)}" else ""}"
                tile.updateTile()
                AudioProfileList.applyProfile(applicationContext)
            } catch(_: Exception) {
                // Do nothing - no profiles found.
            }
        }
    }
}